charms/tensorboards-web-app/src/charm.py

```diff
@@ -1,158 +1,249 @@
+# Copyright 2023 Canonical Ltd.
 # See LICENSE file for licensing details.
 
-import logging
+"""A Juju Charm for Tensorboards Web App."""
 
-from oci_image import OCIImageResource, OCIImageResourceError
+import logging
+from typing import Dict
+from pathlib import Path
+import yaml
+
+from charmed_kubeflow_chisme.exceptions import ErrorWithStatus, GenericCharmRuntimeError
+from charmed_kubeflow_chisme.kubernetes import (
+    KubernetesResourceHandler,
+    create_charm_default_labels,
+)
+from charmed_kubeflow_chisme.lightkube.batch import delete_many
+from charmed_kubeflow_chisme.pebble import update_layer
+from charms.observability_libs.v1.kubernetes_service_patch import KubernetesServicePatch
+from lightkube import ApiError
+from lightkube.generic_resource import load_in_cluster_generic_resources
+from lightkube.models.core_v1 import ServicePort
+from lightkube.resources.rbac_authorization_v1 import ClusterRole, ClusterRoleBinding
+from lightkube.resources.core_v1 import ServiceAccount
 from ops.charm import CharmBase
 from ops.main import main
-from ops.model import ActiveStatus, BlockedStatus, MaintenanceStatus, WaitingStatus
+from ops.model import ActiveStatus, BlockedStatus, Container, MaintenanceStatus, WaitingStatus
+from ops.pebble import Layer
 from serialized_data_interface import NoCompatibleVersions, NoVersionsListed, get_interfaces
 
-
-class CheckFailed(Exception):
-    """Raise this exception if one of the checks in main fails."""
-
-    def __init__(self, msg: str, status_type=None):
-        super().__init__()
-
-        self.msg = str(msg)
-        self.status_type = status_type
-        self.status = status_type(self.msg)
+METADATA = yaml.safe_load(Path("./metadata.yaml").read_text())
+K8S_RESOURCES = {
+    "template_files": ["src/templates/auth_manifests.yaml.j2"],
+    "resource_types": {ClusterRole, ClusterRoleBinding, ServiceAccount},
+    "scope": "tensorboard",
+}
+LIGHTKUBE_FIELD_MANAGER = "lightkube"
+PORT = 5000
 
 
-class Operator(CharmBase):
+class TensorboardsWebApp(CharmBase):
     def __init__(self, *args):
+        """Initialize charm and setup the container."""
         super().__init__(*args)
 
-        self.log = logging.getLogger(__name__)
-        self.image = OCIImageResource(self, "oci-image")
-
-        for event in [
-            self.on.install,
-            self.on.upgrade_charm,
-            self.on.config_changed,
-            self.on.leader_elected,
-            self.on["ingress"].relation_changed,
-        ]:
-            self.framework.observe(event, self.main)
-
-    def main(self, event):
-        try:
-
-            self._check_leader()
+        self.logger = logging.getLogger(__name__)
+        self._namespace = self.model.name
+        self._name = self.app.name
 
-            interfaces = self._get_interfaces()
+        # Explicitly define container name since the charm can be deployed with different app name
+        self._container_name = list(METADATA["containers"])[0]
```

Comment:
Hmm with that in mind, I 'm not sure if it really makes sense then to parse the `metadata.yaml` for the container name, since doing `METADATA["containers"]["<your container name>"]` will return the field's value but what we really need is the field's key (`<your container name>`).

Reply:
that's true, then nothing to do here.